File: Pharmaceutical_Company_Example/02_Structural/Java/Composite.java
// Component: Abstract class for Regulatory Process
abstract class RegulatoryProcess {
    public abstract void execute();
}

// Leaf: Represents individual steps in the regulatory process
class Step extends RegulatoryProcess {
    private String stepName;

    public Step(String stepName) {
        this.stepName = stepName;
    }

    @Override
    public void execute() {
        System.out.println("Executing step: " + stepName);
    }
}

// Composite: Represents composite processes made up of steps or other processes
class CompositeProcess extends RegulatoryProcess {
    private String processName;
    private List<RegulatoryProcess> children = new ArrayList<>();

    public CompositeProcess(String processName) {
        this.processName = processName;
    }

    public void add(RegulatoryProcess process) {
        children.add(process);
    }

    public void remove(RegulatoryProcess process) {
        children.remove(process);
    }

    @Override
    public void execute() {
        System.out.println("Executing composite process: " + processName);
        for (RegulatoryProcess child : children) {
            child.execute();
        }
    }
}

// Client code
public class Main {
    public static void main(String[] args) {
        // Leaf objects - individual steps
        RegulatoryProcess clinicalTrialStep = new Step("Clinical Trial Approval");
        RegulatoryProcess safetyStep = new Step("Safety Review");
        RegulatoryProcess manufacturingStep = new Step("Manufacturing Validation");

        // Composite processes
        CompositeProcess clinicalProcess = new CompositeProcess("Clinical Process");
        clinicalProcess.add(clinicalTrialStep);

        CompositeProcess regulatoryProcess = new CompositeProcess("Regulatory Process");
        regulatoryProcess.add(safetyStep);
        regulatoryProcess.add(manufacturingStep);

        CompositeProcess fullApprovalProcess = new CompositeProcess("Full Drug Approval Process");
        fullApprovalProcess.add(clinicalProcess);
        fullApprovalProcess.add(regulatoryProcess);

        // Execute the full approval process
        fullApprovalProcess.execute();
    }
}

File: Pharmaceutical_Company_Example/02_Structural/PHP/Composite.php
<?php

// Component: RegulatoryProcess
abstract class RegulatoryProcess {
    abstract public function execute(): void;
}

// Leaf: Step (Individual Steps in the Process)
class Step extends RegulatoryProcess {
    private string $stepName;

    public function __construct(string $stepName) {
        $this->stepName = $stepName;
    }

    public function execute(): void {
        echo "Executing step: {$this->stepName}\n";
    }
}

// Composite: CompositeProcess (A Collection of Steps)
class CompositeProcess extends RegulatoryProcess {
    private string $processName;
    private array $children = [];

    public function __construct(string $processName) {
        $this->processName = $processName;
    }

    public function add(RegulatoryProcess $regulatoryProcess): void {
        $this->children[] = $regulatoryProcess;
    }

    public function remove(RegulatoryProcess $regulatoryProcess): void {
        $index = array_search($regulatoryProcess, $this->children, true);
        if ($index !== false) {
            array_splice($this->children, $index, 1);
        }
    }

    public function execute(): void {
        echo "Executing composite process: {$this->processName}\n";
        foreach ($this->children as $child) {
            $child->execute();
        }
    }
}

// Leaf objects - individual steps
$clinicalTrialStep = new Step("Clinical Trial Approval");
$safetyStep = new Step("Safety Review");
$manufacturingStep = new Step("Manufacturing Validation");

// Composite processes
$clinicalProcess = new CompositeProcess("Clinical Process");
$clinicalProcess->add($clinicalTrialStep);

$regulatoryProcess = new CompositeProcess("Regulatory Process");
$regulatoryProcess->add($safetyStep);
$regulatoryProcess->add($manufacturingStep);

$fullApprovalProcess = new CompositeProcess("Full Drug Approval Process");
$fullApprovalProcess->add($clinicalProcess);
$fullApprovalProcess->add($regulatoryProcess);

$fullApprovalProcess->execute();

?>
